Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportStreamCocoa.mm

```diff
@@ -41,46 +47,162 @@
 {
-    ASSERT(m_streamType != NetworkTransportStreamType::IncomingUnidirectional);
-    nw_connection_send(m_connection.get(), makeDispatchData(Vector(data)).get(), NW_CONNECTION_DEFAULT_MESSAGE_CONTEXT, withFin, ^(nw_error_t error) {
-        // FIXME: Pipe any error to JS.
-        // FIXME: sendBytes should probably have a completion handler.
-    });
+    if (m_streamState == NetworkTransportStreamState::WriteClosed) {
+        completionHandler(WebCore::Exception(WebCore::ExceptionCode::InvalidStateError));
+        return;
+    }
+    nw_connection_send(m_connection.get(), makeDispatchData(Vector(data)).get(), NW_CONNECTION_DEFAULT_MESSAGE_CONTEXT, withFin, makeBlockPtr([weakThis = WeakPtr { *this }, withFin = withFin, completionHandler = WTFMove(completionHandler)] (nw_error_t error) mutable {
+        RefPtr strongThis = weakThis.get();
+        if (!strongThis)
+            return;
+        if (error) {
+            if (nw_error_get_error_domain(error) == nw_error_domain_posix && nw_error_get_error_code(error) == ECANCELED) {
+                // If we were cancelled, exit gracefully
+                completionHandler(std::nullopt);
+            } else
+                completionHandler(WebCore::Exception(WebCore::ExceptionCode::NetworkError));
+            return;
+        }
+
+        completionHandler(std::nullopt);
+
+        if (withFin) {
+            switch (strongThis->m_streamState) {
+            case NetworkTransportStreamState::Ready:
+                strongThis->m_streamState = NetworkTransportStreamState::WriteClosed;
+                break;
+            case NetworkTransportStreamState::ReadClosed:
+                strongThis->cancelSend(std::nullopt);
+                break;
+            case NetworkTransportStreamState::WriteClosed:
+                RELEASE_ASSERT_NOT_REACHED();
+            }
+        }
+    }).get());
 }
 
 void NetworkTransportStream::receiveLoop()
 {
-    ASSERT(m_streamType != NetworkTransportStreamType::OutgoingUnidirectional);
+    if (m_streamState == NetworkTransportStreamState::ReadClosed) {
+        RELEASE_ASSERT_NOT_REACHED();
+        return;
+    }
     nw_connection_receive(m_connection.get(), 0, std::numeric_limits<uint32_t>::max(), makeBlockPtr([weakThis = WeakPtr { *this }] (dispatch_data_t content, nw_content_context_t, bool withFin, nw_error_t error) {
         RefPtr strongThis = weakThis.get();
         if (!strongThis)
             return;
-        if (error)
-            return; // FIXME: Pipe this error to JS.
-        if (!content && withFin)
+        RefPtr session = strongThis->m_session.get();
+        if (!session)
+            return;
+        if (error) {
+            if (!(nw_error_get_error_domain(error) == nw_error_domain_posix && nw_error_get_error_code(error) == ECANCELED))
+                session->streamReceiveBytes(strongThis->m_identifier, { }, false, WebCore::Exception(WebCore::ExceptionCode::NetworkError));
+            // If we were cancelled, exit gracefully
```

Comment:
Do we want to do anything with the error if it is not ECANCELED?  This comment is a little mysterious.